llvm/lib/CodeGen/PrologEpilogInserter.cpp

```diff
@@ -923,20 +911,28 @@ void PEIImpl::calculateFrameObjectOffsets(MachineFunction &MF) {
   Align MaxAlign = MFI.getMaxAlign();
   // First assign frame offsets to stack objects that are used to spill
   // callee saved registers.
-  if (MaxCSFrameIndex >= MinCSFrameIndex) {
-    for (unsigned i = 0; i <= MaxCSFrameIndex - MinCSFrameIndex; ++i) {
-      unsigned FrameIndex =
-          StackGrowsDown ? MinCSFrameIndex + i : MaxCSFrameIndex - i;
-
+  if (StackGrowsDown) {
```

Comment:
Feel free to reject this, but I think you could do something like:
```c++
  auto AllFIsSeq = seq(MFI.getObjectIndexBegin(), MFI.getObjectIndexEnd());
  for (auto [FILo, FIHi] : zip_equal(AllFIsSeq, reverse(AllFIsSeq))) {
    int FI = StackGrowsDown ? FILo : FIHi;
    // Only allocate objects on the default stack.
    if (!MFI.isCalleeSavedObjectIndex(FI) ||
        MFI.getStackID(FI) != TargetStackID::Default)
      continue;

    // TODO: should this just be if (MFI.isDeadObjectIndex(FI))
    if (!StackGrowsDown && MFI.isDeadObjectIndex(FI))
      continue;

    AdjustStackOffset(MFI, FI, StackGrowsDown, Offset, MaxAlign);
  }
```
Rather than duplicate the loop for forwards/reverse (IIUC).

Reply:
Personally, I'd find that *much* harder to reason about.  The idea of using a sequence is interesting though, maybe we could conditional reverse a sequence instead?  I'll play with this a bit and may tag you for review on a followup.

Reply:
No worries :+1: You could wrap this trick up as:
```c++
template <typename ContainerTy>
[[nodiscard]] auto conditionally_reverse(ContainerTy &&C, bool Reverse) {
  return map_range(zip_equal(reverse(C), C), [Reverse](auto I) {
    return Reverse ? std::get<0>(I) : std::get<1>(I);
  });
}
```

(Probably would fit in `llvm/ADT/STLExtras.h`)

Which allows for:
```c++
auto AllFIs = seq(MFI.getObjectIndexBegin(), MFI.getObjectIndexEnd());
for (auto FI : conditionally_reverse(AllFIs, /*Reverse=*/!StackGrowsDown)) {
  // ...
}
```